p2p/transport/quicreuse/reuse.go

```diff
@@ -203,29 +222,50 @@ func (r *reuse) dialLocked(network string, source *net.IP) (*reuseConn, error) {
 		return nil, err
 	}
 	rconn := newReuseConn(conn)
-	r.global[conn.LocalAddr().(*net.UDPAddr).Port] = rconn
+	r.globalDialers[conn.LocalAddr().(*net.UDPAddr).Port] = rconn
 	return rconn, nil
 }
 
 func (r *reuse) Listen(network string, laddr *net.UDPAddr) (*reuseConn, error) {
-	conn, err := net.ListenUDP(network, laddr)
-	if err != nil {
-		return nil, err
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	var rconn *reuseConn
+	var localAddr *net.UDPAddr
+
+	// reuse the fallback connection if we've dialed out from this port already
+	if laddr.IP.IsUnspecified() {
+		if laddr.Port == 0 {
+			// use any connection we have dialed out of
+			for _, conn := range r.globalDialers {
+				rconn = conn
+				localAddr = rconn.UDPConn.LocalAddr().(*net.UDPAddr)
+				delete(r.globalDialers, localAddr.Port)
```

Comment:
Add a comment that it will be added to the `globalListeners` map below.

Reply:
does this look fine?

Reply:
Actually, I think it would be clearer if you added it to the `globalListeners` map here and returned early.

Reply:
I agree. that's much better.

Reply:
fixed.